Share graphics context globally

psychon opened the pull request:
Instead of creating a graphics context for every surface_t, this commit
adds a cache that allows to "remember" up to two GCs. Thus, the code
uses less GCs. When a GC from the cache can be used, this also gets rid
of a round-trip to the X11 server. Both of these are tiny, insignificant
savings, but so what?

Fixes: https://github.com/i3/i3/issues/3478
Signed-off-by: Uli Schlachter <psychon@znc.in>

-----

This adds a copy&paste duplicate of `get_visual_depth()` from `src/xcb.c` to `libi3. If you tell me what to do, I can clean this up again, but I was not really sure what the best option here is.

I tried testing the cache overflow by trying to create a window with a depth other than 24 or 32, but my X11 server does not offer any such visuals:
```
$ xdpyinfo| grep depth: | uniq
    depth:    24 planes
    depth:    32 planes
```
Also, initially I planned to use a `SLIST` for this and have the cache unlimited. After briefly staring at the macros, I went with a statically sized solution instead. Thus, there is now the possibility that a GC cannot be cached and is thus still owned by the `surface_t`.

psychon committed:
Share graphics context globally

Instead of creating a graphics context for every surface_t, this commit
adds a cache that allows to "remember" up to two GCs. Thus, the code
uses less GCs. When a GC from the cache can be used, this also gets rid
of a round-trip to the X11 server. Both of these are tiny, insignificant
savings, but so what?

Since GCs are per-depth, this code needs access to get_visual_depth().
To avoid a code duplication, this function is moved to libi3.

Fixes: https://github.com/i3/i3/issues/3478
Signed-off-by: Uli Schlachter <psychon@znc.in>

orestisfl commented on libi3/draw_util.c:
```diff
@@ -58,7 +116,9 @@ void draw_util_surface_init(xcb_connection_t *conn, surface_t *surface, xcb_draw
 void draw_util_surface_free(xcb_connection_t *conn, surface_t *surface) {
-    xcb_free_gc(conn, surface->gc);
+    if (surface->owns_gc) {
+        xcb_free_gc(conn, surface->gc);
```
What happens after you free the gc that is used in other surfaces?

psychon replied:
Uhm... Sorry, but I don't quite get the question...?

I just re-read the diff and `owns_gc` is only set by `get_gc()`. That function only ever sets `owns_gc = true` when nothing else references the GC. Put differently: If the resulting GC is saved in the cache (either already before or during the call to `get_gc()`), `owns_gc = false`. If the resulting GC is only referenced by the `draw_util_surface`, then `owns_gc = true`.

So, to try to answer the question: If this were to free a GC that is also used in other `draw_util_surface`s, that would be bad. But I don't think that can possibly happen.

orestisfl replied:
Correct, never mind my comment

orestisfl commented on libi3/draw_util.c:
```diff
@@ -28,6 +35,64 @@ static void draw_util_set_source_color(surface_t *surface, color_t color);
+    return gc;
+}
+
+static uint16_t get_visual_depth(xcb_visualid_t visual_id) {
```
I would delete this function from xcb.c and move it to libi3 to not have duplicated code

orestisfl commented on libi3/draw_util.c:
```diff
@@ -20,6 +20,13 @@ extern xcb_visualtype_t *visual_type;
+static struct {
+    uint8_t depth;
+    xcb_gcontext_t gc;
+} gc_cache[2] = {
```
I feel like in i3 we are sticking too much in primitive data structures even though we can take advantage of more advanced ones, e.g. this could have been implemented more easily with [glib's hashtable](https://docs.gtk.org/glib/struct.HashTable.html) and glib is already a dependency.

I am not really requesting that you re-write this, just an observation. You can ignore this comment.

psychon replied:
Yeah, I didn't use much glib so far. I quickly looked at the link and then decided to go with `You can ignore this comment`. :sweat_smile:

orestisfl commented on libi3/draw_util.c:
```diff
@@ -20,6 +20,13 @@ extern xcb_visualtype_t *visual_type;
 /* Forward declarations */
 static void draw_util_set_source_color(surface_t *surface, color_t color);
 
+static struct {
```
This static struct can live entirely inside the `get_gc` function since it's only referenced there.

orestisfl reviewed (changes requested):
Thank you, and sorry for the delayed review

psychon commented:
I rebased this on `next` (had some conflicts) and addressed the comments.

I am not entirely sure, but the following comments now seems outdated:
```c
    if (surface->owns_gc) {
        /* NOTE: This function is also called on uninitialised surface_t instances.
         * The x11 error from xcb_free_gc(conn, XCB_NONE) is silently ignored
         * elsewhere.
         */
        xcb_free_gc(conn, surface->gc);
    }
```
`uninitialised` cannot really mean "contains arbitrary garbage". That sounds like a recipe for crashes. So I guess this means zero-initialised. In that case, `owns_gc` should be `false` and I just made this error case go away.

However, this is just a guess and I don't have the time right now to dig deeper, so I am just mentioning this here and keeping things otherwise the way they already were.

orestisfl commented on libi3/draw_util.c:
```diff
@@ -68,11 +130,13 @@ void draw_util_surface_free(xcb_connection_t *conn, surface_t *surface) {
+        /* NOTE: This function is also called on uninitialised surface_t instances.
+         * The x11 error from xcb_free_gc(conn, XCB_NONE) is silently ignored
+         * elsewhere.
+         */
```
> I rebased this on `next` (had some conflicts) and addressed the comments.
> 
> I am not entirely sure, but the following comments now seems outdated:
> 
> ```c
>     if (surface->owns_gc) {
>         /* NOTE: This function is also called on uninitialised surface_t instances.
>          * The x11 error from xcb_free_gc(conn, XCB_NONE) is silently ignored
>          * elsewhere.
>          */
>         xcb_free_gc(conn, surface->gc);
>     }
> ```
> 
> `uninitialised` cannot really mean "contains arbitrary garbage". That sounds like a recipe for crashes. So I guess this means zero-initialised. In that case, `owns_gc` should be `false` and I just made this error case go away.
> 
> However, this is just a guess and I don't have the time right now to dig deeper, so I am just mentioning this here and keeping things otherwise the way they already were.

The comment mentions `XCB_NONE` so that's what I imagine it means, not garbage. I've played around and we can still get errors on the free but they happen in `next` as well, we can continue ignoring them for now. I think we can remove the comment as it seems misleading now.

```suggestion
```

orestisfl commented on libi3/draw_util.c:
```diff
@@ -28,6 +28,76 @@ static bool surface_initialized(surface_t *surface) {
+    }
+
+    xcb_gcontext_t gc = xcb_generate_id(conn);
+    xcb_void_cookie_t gc_cookie = xcb_create_gc_checked(conn, gc, drawable, 0, NULL);
```
```suggestion
    /* The drawable is only used to get the root and depth, thus the gc is not
     * tied to the drawable and it can be re-used from different drawables. */
    xcb_void_cookie_t gc_cookie = xcb_create_gc_checked(conn, gc, drawable, 0, NULL);
```

I came up with this while trying to understand why this is safe to do e.g. when creating a gc from a drawable that is later freed. If my comment is correct, maybe we can include it for further context to readers?

psychon replied:
Done.

"Definitive answer" seems to come from the X11 protocol reference manual: https://www.x.org/releases/X11R7.6/doc/xproto/x11protocol.html#requests:CreateGC

>  The gcontext can be used with any destination drawable having the same root and depth as the specified drawable [...]

orestisfl reviewed (changes requested):
Thanks, couple of last things

psychon committed:
Add review suggestions from orestisfl

Signed-off-by: Uli Schlachter <psychon@znc.in>

orestisfl reviewed (approved):
Thank you
